File: src/src_core/fmg/common/geom/BoundDouble.java
package fmg.common.geom;

import java.util.Locale;

/** Padding / Margin */
public class BoundDouble {

    public double left, right, top, bottom;

    public BoundDouble(BoundDouble copy) { this.left = copy.left; this.top = copy.top; this.right = copy.right; this.bottom = copy.bottom; }
    public BoundDouble(double bound) { left = top = right = bottom = bound; }
    public BoundDouble(double left, double top, double right, double bottom) {
        this.left   = left;
        this.top    = top;
        this.right  = right;
        this.bottom = bottom;
    }

    public double getLeftAndRight() { return left + right; }
    public double getTopAndBottom() { return top + bottom; }

    public boolean isEmpty() { return left == 0 && right == 0 && top == 0 && bottom == 0; }

    @Override
    public int hashCode() {
        long temp = Double.doubleToLongBits(bottom);
        int result = 31 + (int) (temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(left);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(right);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(top);
        return 31 * result + (int) (temp ^ (temp >>> 32));
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null) return false;
        if (!(obj instanceof BoundDouble))
            return false;
        return equals((BoundDouble) obj);
    }

    public boolean equals(BoundDouble other) {
        return (other != null) &&
               (Double.doubleToLongBits(bottom) == Double.doubleToLongBits(other.bottom)) &&
               (Double.doubleToLongBits(left)   == Double.doubleToLongBits(other.left)) &&
               (Double.doubleToLongBits(right)  == Double.doubleToLongBits(other.right)) &&
               (Double.doubleToLongBits(top)    == Double.doubleToLongBits(other.top));
    }

    @Override
    public String toString() { return String.format(Locale.US, "{lft=%.2f, rght=%.2f, top=%.2f, bttm=%.2f}", left, right, top, bottom); }

}

File: FastMines_core/src/fmg/common/geom/BoundDouble.cs
namespace Fmg.Common.Geom {

    /// <summary> Padding / Margin </summary>
    public struct BoundDouble {

        public BoundDouble(BoundDouble copy) { Left = copy.Left; Top = copy.Top; Right = copy.Right; Bottom = copy.Bottom; }
        public BoundDouble(double left, double top, double right, double bottom) {
            Left   = left;
            Top    = top;
            Right  = right;
            Bottom = bottom;
        }

        public BoundDouble(double bound) {
            Left = Top = Right = Bottom = bound;
        }

        public double Left   { get; set; }
        public double Right  { get; set; }
        public double Top    { get; set; }
        public double Bottom { get; set; }

        public double LeftAndRight => Left + Right;
        public double TopAndBottom => Top + Bottom;

        public SizeDouble LeftTopOffset => new SizeDouble(Left, Top);

        public bool Equals(BoundDouble other) {
            return this == other;
        }

        public override bool Equals(object other) {
            if (ReferenceEquals(null, other))
                return false;
            return (other is BoundDouble) && (this == (BoundDouble)other);
        }

        public override int GetHashCode() {
            unchecked {
                var hashCode = Left.GetHashCode();
                hashCode = (hashCode * 397) ^ Right.GetHashCode();
                hashCode = (hashCode * 397) ^ Top.GetHashCode();
                hashCode = (hashCode * 397) ^ Bottom.GetHashCode();
                return hashCode;
            }
        }

        public override string ToString() {
            return string.Format("{{Left:{0:0.00}, Top:{1:0.00}, Right:{2:0.00}, Bottom:{3:0.00}}}", Left, Top, Right, Bottom);
        }


        public static bool operator ==(BoundDouble t1, BoundDouble t2) {
            return t1.Left.HasMinDiff(t2.Left) && t1.Top.HasMinDiff(t2.Top) && t1.Right.HasMinDiff(t2.Right) && t1.Bottom.HasMinDiff(t2.Bottom);
        }

        public static bool operator !=(BoundDouble t1, BoundDouble t2) {
            return !t1.Left.HasMinDiff(t2.Left) || !t1.Top.HasMinDiff(t2.Top) || !t1.Right.HasMinDiff(t2.Right) || !t1.Bottom.HasMinDiff(t2.Bottom);
        }

    }

}
